### DISCUSSION TITLE
Single-value tag widgets

### DISCUSSION THREAD

**Discussion by lgimberis** [codingjoe/django-select2#126] [Q&A]
Hi, firstly thanks for your work on this project. It's exactly what I was looking for.

I'm currently trying to implement a select2 field that allows the end user to either select an existing value OR enter a 'new' value, which will automatically create the related object. It seems that the ModelSelect2TagWidget was made for exactly this purpose.
However, this widget inherently allows selecting multiple items at once. In the context where I'm using it, this isn't appropriate and the user should only be able to enter a single value.

Is there a straightforward way to configure a ModelSelect2TagWidget to only select a single item, or ModelSelect2Widget to allow 'tagging'/submission of items that don't yet exist? 
To clarify, something more like the first example on [this page](https://select2.org/tagging), currently it's the second.
Right now, the best way I can see to achieve this would be to make a custom set of widgets parallel to the existing tag widgets but without the multiple mixin - is there a simpler way?

Thanks.

**Comment by codingjoe** [comment]
Hi @lgimberis,

Thanks for reaching out. This is a somewhat unique use case, but not impossible. If you have a look at the inheritance graph here, you can see how the tag widget inherits from a multiple select input. https://django-select2.readthedocs.io/en/latest/django_select2.html#module-django_select2.forms

However, if you look at the corresponding code, it should be fairly simple to recreate the same for a single select input.

If you need further assistance, I can go into more detail.

Best Joe

**Comment by rez0n** [comment]
Hi, I wonder to share my solution for a single tag selection/creation widget.

```
def represent_int(s):
    try:
        int(s)
        return True
    except ValueError:
        return False


class SingleTagPickCreateWidget(s2forms.ModelSelect2TagWidget):
    allow_multiple_selected = False
    queryset = Tag.objects.all()
    search_fields = [
        "name__icontains",
    ]
    
    def build_attrs(self, base_attrs, extra_attrs=None):
        """Add select2's tag attributes."""
        default_attrs = {
            "data-minimum-input-length": 1,
            "data-tags": "true",
            "data-token-separators": '[]',
        }
        default_attrs.update(base_attrs)
        return super().build_attrs(default_attrs, extra_attrs=extra_attrs)

    def value_from_datadict(self, data, files, name):
        # Get input value as list
        values = super().value_from_datadict(data, files, name)
        queryset = self.get_queryset()
        # Get database object pk if input value isdigits (ie. selected existing object)
        pks = queryset.filter(**{'pk__in': [v for v in values if v.isdigit()]}).values_list('pk', flat=True)
        for val in values:
            if represent_int(val) and int(val) not in pks or not represent_int(val) and force_str(val) not in pks:
                if not queryset.filter(name=val).exists():
                    val = queryset.create(name=val).pk
                else:
                    val = queryset.get(name=val).pk
        return val
```
